File: tests/rust/game/counter.rs
//! The bystander: advances every tick and records the dt it read
//! through `labelle::dt()` — proving the Controller's stamp reached the
//! contract before updates ran, and that siblings keep running whatever
//! the scripts around it do.

use crate::labelle::{self, EntityId, Script};

#[derive(Default)]
pub struct Counter {
    e: EntityId,
    n: u32,
}

impl Script for Counter {
    fn init(&mut self) {
        self.e = labelle::create_entity();
    }

    fn update(&mut self, dt: f32) {
        // The stamped dt and the passed dt are the same tick value.
        assert_eq!(labelle::dt(), dt, "dt stamp skew");
        self.n += 1;
        // Keys sorted (dt < n) — the suites pin component JSON
        // byte-for-byte, matching the embedded preludes' sorted encoders.
        labelle::set_component(
            self.e,
            "Counter",
            &format!("{{\"dt\":{},\"n\":{}}}", dt, self.n),
        );
    }
}

File: tests/go/game/counter.go
// A healthy bystander: counts ticks into a Counter component. Entity
// creation happens in Init, so registration order = entity id order —
// the suites lean on that.
package game

import (
	"fmt"

	"labelle"
)

type Counter struct {
	labelle.NoopScript
	entity labelle.EntityID
	n      int
}

func (c *Counter) Init() {
	c.entity = labelle.CreateEntity()
}

func (c *Counter) Update(dt float32) {
	c.n++
	labelle.SetComponent(c.entity, "Counter",
		fmt.Sprintf(`{"dt":%v,"n":%d}`, dt, c.n))
}
